libr/anal/sign.c

```diff
@@ -1,235 +1,357 @@
-		if (!(data = r_anal_strmask (anal, arg))) {
-			r_sign_item_free (si);
-			si = NULL;
+	// Deserialize val: <type>,size,bytes,mask
+	for (ptr = (char*) v2, i = 0; ; ptr = NULL, i++) {
+		token = strtok (ptr, "|");
+		if (!token) {
 			break;
 		}
-		len = strlen (data) + 4; // \xf0
-		si->bytes = (ut8 *)calloc (R_MAX (len, 4), 1);
-		si->mask = (ut8 *)calloc (R_MAX (len, 4), 1);
-		if (!si->bytes || !si->mask) {
-			eprintf ("Cannot malloc\n");
-			r_sign_item_free (si);
-			si = NULL;
+
+		switch (i) {
+		case 0:
+			it->type = token[0];
 			break;
+		case 1:
+			it->size = sdb_atoi (token);
+			break;
+		case 2:
+			it->bytes = malloc (it->size);
+			r_hex_str2bin(token, it->bytes);
+			break;
+		case 3:
+			it->mask = malloc (it->size);
+			r_hex_str2bin(token, it->mask);
+			break;
+		default:
+			retval = false;
+			goto exit_function;
 		}
-		si->size = r_hex_str2binmask (data, si->bytes, si->mask);
-		if (si->size < 1) {
-			r_sign_item_free (si);
-			si = NULL;
+	}
+
+exit_function:
+	free (k2);
+	free (v2);
+
+	return retval;
+}
+
+static void serialize(RSignItem *it, char *k, char *v) {
+	char *hexbytes = NULL, *hexmask = NULL;
+	int len;
+
+	if (k) {
+		snprintf (k, R_SIGN_KEY_MAXSZ, "zign|%d|%s", it->space, it->name);
+	}
+
+	if (v) {
+		len = it->size * 2 + 1;
+
+		hexbytes = calloc (1, len);
+		hexmask = calloc (1, len);
+
+		r_hex_bin2str (it->bytes, it->size, hexbytes);
+		r_hex_bin2str (it->mask, it->size, hexmask);
+
+		snprintf (v, R_SIGN_VAL_MAXSZ, "%c|%d|%s|%s", it->type, it->size, hexbytes, hexmask);
+
+		free (hexbytes);
+		free (hexmask);
+	}
+}
+
+R_API bool r_sign_add(RAnal *a, int type, const char *name, ut64 size, const ut8 *bytes, const ut8 *mask) {
+	RSignItem *it = R_NEW0 (RSignItem);
+	char key[R_SIGN_KEY_MAXSZ], val[R_SIGN_VAL_MAXSZ];
+	bool retval = true;
+
+	it->type = type;
+	it->name = r_str_new (name);
+	it->space = a->zign_spaces.space_idx;
+	it->size = size;
+
+	it->bytes = malloc (size);
+	memcpy (it->bytes, bytes, size);
+	it->mask = malloc (size);
+	memcpy (it->mask, mask, size);
+
+	serialize (it, key, val);
+
+	if (sdb_exists (a->sdb_zigns, key)) {
+		eprintf ("error: zignature already exists\n");
+		retval = false;
+		goto exit_function;
+	}
+
+	sdb_set (a->sdb_zigns, key, val, 0);
+
+exit_function:
+	r_sign_item_free (it);
+
+	return retval;
+}
+
+R_API bool r_sign_add_anal(RAnal *a, const char *name, ut64 size, const ut8 *bytes) {
+	ut8 *mask = NULL;
+	bool retval = true;
+
+	mask = r_anal_mask (a, size, bytes);
+	retval = r_sign_add(a, R_SIGN_ANAL, name, size, bytes, mask);
+
+	free (mask);
+	return retval;
+}
+
+struct ctxDeleteCB {
+	char buf[R_SIGN_KEY_MAXSZ];
+	RAnal *anal;
+};
+
+int zignDeleteCB(void *user, const char *k, const char *v) {
+	struct ctxDeleteCB *ctx = (struct ctxDeleteCB *)user;
+
+	if (r_str_cmp (k, ctx->buf, strlen (ctx->buf))) {
+		sdb_remove (ctx->anal->sdb_zigns, k, 0);
+	}
+
+	return 1;
+}
+
+R_API bool r_sign_delete(RAnal *a, const char *name) {
```

Comment:
Actually it is guaranteed. The string shouldn't be changed from this function because the name is only used to query sdb.